From aea031169389b458e679c9be813f75f100b159d5 Mon Sep 17 00:00:00 2001
From: Ash Charles <ashcharles@gmail.com>
Date: Wed, 15 Oct 2014 12:04:25 -0700
Subject: [PATCH 16/16] mwifiex: Specify single-thread work queue

On the Gumstix DuoVero platform, the mwifiex_sdio driver will, after
either heavy usage or several hours of idle time, eventually start
spewing fault messages like this:

    Modules linked in: mwifiex_sdio btmrvl_sdio mwifiex btmrvl bluetooth cfg80211 firmware_class rfkill ipv6
    [<c0014c84>] (unwind_backtrace+0x0/0x11c) from [<c048dbbc>] (__schedule_bug+0x48/0x60)
    [<c048dbbc>] (__schedule_bug+0x48/0x60) from [<c0493bbc>] (__schedule+0x68/0x798)
    [<c0493bbc>] (__schedule+0x68/0x798) from [<c000f0b0>] (cpu_idle+0x104/0x120)
    [<c000f0b0>] (cpu_idle+0x104/0x120) from [<80489c54>] (0x80489c54)
    BUG: scheduling while atomic: swapper/1/0/0xfffffe00

If the 'MWIFIEX_WORK_QUEUE' is created as a 'singlethread' workqueue,
this error doesn't appear.  Note that the DuoVero is a dual-core
machine.

Signed-off-by: Ash Charles <ashcharles@gmail.com>
---
 drivers/net/wireless/mwifiex/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/mwifiex/main.c b/drivers/net/wireless/mwifiex/main.c
index 7e8031b..4f9914f 100644
--- a/drivers/net/wireless/mwifiex/main.c
+++ b/drivers/net/wireless/mwifiex/main.c
@@ -728,7 +728,7 @@ mwifiex_add_card(void *card, struct semaphore *sem,
 	adapter->cmd_wait_q.status = 0;
 	adapter->scan_wait_q_woken = false;
 
-	adapter->workqueue = create_workqueue("MWIFIEX_WORK_QUEUE");
+	adapter->workqueue = create_singlethread_workqueue("MWIFIEX_WORK_QUEUE");
 	if (!adapter->workqueue)
 		goto err_kmalloc;
 
-- 
1.8.3.2

